File: LeetCode/003. Longest Substring Without Repeating Characters/solution.py
from __future__ import annotations


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        dp = 0
        mp = {}
        for i, ch in enumerate(s):
            if ch in mp and mp[ch]+dp>=i:
                dp = i-mp[ch]
            else:
                dp += 1
            mp[ch] = i
            ans = max(ans, dp)
        return ans

s = Solution()
sr = "abcabcbb"

s.lengthOfLongestSubstring(sr)


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dp, res, n = 0, 0, len(s)
        last = [-1]*256
        for i in range(n):
            dp = max(dp, last[ord(s[i])]+1)
            res = max(res, i-dp+1)
            last[ord(s[i])] = i
        return res

File: LeetCode/003. Longest Substring Without Repeating Characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int dp = 0, res = 0, n = s.length();
        vector<int> last(256,-1);
        for(int i=0;i<n;i++){
            int ch = s[i];
            dp = max(dp, last[ch]+1);
            res = max(res, i-dp+1);
            last[ch] = i;
        }
        return res;
    }
};


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> cnt(256, 0);
        int n = s.length();
        int left = 0, res = 0;
        for (int right = 0; right < n; right++){
            cnt[s[right]]++;
            while (cnt[s[right]] > 1)
                cnt[s[left++]]--;
            res = max(res, right-left+1);
        }
        return res;
    }
};
